# Uittesten van strategie elementen in Pandas voor CryptoMaven bot. 

## Swing trade strategie
De bedoeling is dat deze strategie signalen geeft op de volgende indicatoren:
* DEM9 > SMA 21 
* RSI14 > (SMA14 over RSI14)
* MFI 14 > (SMA21 over MFI14)

Deze strategie is een mogelijke verbetering op de strategie ``DEMA 9 > SMA 21 || RSI21 > 55`` die voor Bitcoin 281% en Ethereum 292% winst gaf. Het geeft iig meer confirmation op een ingeslagen richting (tbv. mijn psychologie...)

## RSI strategie
Tevens moet het dataframe signalen geven op de momenten waarop de RSI < 35 is. Zeker op 1h timeframe. Dit is een indicatie dat prijs oversold is en opgelet moet worden op signalen waarop ingestapt kan worden. Bij voorkeur na een prijsactie waarbij de prijs lager sluit bij een hogere RSI (bullish divergence).



![CryptoMaven2 Chart](./images/CryptoMaven2_chart.png)

## Data in dataframe laden en controleren op juistheid

Data wordt dmv. een python3 functie gedownload van Binance.

In [2]:
import pandas as pd
import pandas_ta as ta
import numpy as np
import matplotlib.pyplot as plt

# Load data
df = pd.read_csv("./BTCUSDT_1h.csv")

# Reading columns
df.columns

Index(['# openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime',
       'qouteAssetVolume', 'numberOfTrades', 'takerBuyBaseAssetVolume',
       'takerBuyQuoteAssetVolume', 'ignore'],
      dtype='object')

In [3]:
df.shape

(500, 12)

In [4]:
df.info

<bound method DataFrame.info of         # openTime      open      high       low     close       volume  \
0    1603220400000  11964.58  11981.20  11881.00  11912.35  2744.144578   
1    1603224000000  11912.42  11912.79  11855.00  11897.92  2523.731342   
2    1603227600000  11897.92  11937.73  11877.49  11922.53  1409.942909   
3    1603231200000  11922.53  11961.00  11918.22  11950.66  1270.092250   
4    1603234800000  11950.66  11961.16  11885.00  11909.99  1551.691786   
..             ...       ...       ...       ...       ...          ...   
495  1605002400000  15383.84  15430.00  15302.58  15305.14  2365.383922   
496  1605006000000  15305.15  15323.67  15155.00  15279.24  4014.340547   
497  1605009600000  15279.24  15393.65  15213.00  15287.61  3484.423100   
498  1605013200000  15287.61  15330.00  15179.00  15306.83  3370.999279   
499  1605016800000  15306.84  15308.04  15238.01  15239.49   882.070653   

         closeTime  qouteAssetVolume  numberOfTrades  takerBuyBaseA

In [5]:
df.dtypes

# openTime                    int64
open                        float64
high                        float64
low                         float64
close                       float64
volume                      float64
closeTime                     int64
qouteAssetVolume            float64
numberOfTrades                int64
takerBuyBaseAssetVolume     float64
takerBuyQuoteAssetVolume    float64
ignore                        int64
dtype: object

In [6]:
# Aanpassen van unix time naar human time
df['# openTime']=(pd.to_datetime(df['# openTime'],unit='ms'))
df['closeTime']=(pd.to_datetime(df['closeTime'],unit='ms'))

In [7]:
# Watch last 10 entries
df[['# openTime', 'open', 'high', 'low', 'close', 'volume']].tail(10)

# openTime      open      high       low     close       volume
490 2020-11-10 05:00:00  15387.74  15441.18  15308.21  15357.92  2178.822453
491 2020-11-10 06:00:00  15357.92  15390.18  15247.21  15275.82  2558.920212
492 2020-11-10 07:00:00  15275.83  15333.49  15250.47  15274.74  1981.500383
493 2020-11-10 08:00:00  15274.75  15460.00  15261.01  15371.37  3795.218353
494 2020-11-10 09:00:00  15371.36  15440.00  15367.40  15383.84  2253.766369
495 2020-11-10 10:00:00  15383.84  15430.00  15302.58  15305.14  2365.383922
496 2020-11-10 11:00:00  15305.15  15323.67  15155.00  15279.24  4014.340547
497 2020-11-10 12:00:00  15279.24  15393.65  15213.00  15287.61  3484.423100
498 2020-11-10 13:00:00  15287.61  15330.00  15179.00  15306.83  3370.999279
499 2020-11-10 14:00:00  15306.84  15308.04  15238.01  15239.49   882.070653

# Pandas_ta indicatoren toevoegen

In [8]:
df.ta.dema(length=9, append=True)
df.ta.sma(length=21, append=True)
df.ta.rsi(length=14, append=True)
df.ta.mfi(length=14, append=True)

df.columns

Index(['# openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime',
       'qouteAssetVolume', 'numberOfTrades', 'takerBuyBaseAssetVolume',
       'takerBuyQuoteAssetVolume', 'ignore', 'DEMA_9', 'SMA_21', 'RSI_14',
       'MFI_14'],
      dtype='object')

In [9]:
# waardes afronden op 2 cijfers achter de komma
df.round(2)

# openTime      open      high       low     close   volume  \
0   2020-10-20 19:00:00  11964.58  11981.20  11881.00  11912.35  2744.14   
1   2020-10-20 20:00:00  11912.42  11912.79  11855.00  11897.92  2523.73   
2   2020-10-20 21:00:00  11897.92  11937.73  11877.49  11922.53  1409.94   
3   2020-10-20 22:00:00  11922.53  11961.00  11918.22  11950.66  1270.09   
4   2020-10-20 23:00:00  11950.66  11961.16  11885.00  11909.99  1551.69   
..                  ...       ...       ...       ...       ...      ...   
495 2020-11-10 10:00:00  15383.84  15430.00  15302.58  15305.14  2365.38   
496 2020-11-10 11:00:00  15305.15  15323.67  15155.00  15279.24  4014.34   
497 2020-11-10 12:00:00  15279.24  15393.65  15213.00  15287.61  3484.42   
498 2020-11-10 13:00:00  15287.61  15330.00  15179.00  15306.83  3371.00   
499 2020-11-10 14:00:00  15306.84  15308.04  15238.01  15239.49   882.07   

                  closeTime  qouteAssetVolume  numberOfTrades  \
0   2020-10-20 19:59:59.999       32742666.69           39342   
1   2020-10-20 20:59:59.999       29989751.46           39583   
2   2020-10-20 21:59:59.999       16790213.05           28867   
3   2020-10-20 22:59:59.999       15160673.38           27532   
4   2020-10-20 23:59:59.999       18500404.33           24948   
..                      ...               ...             ...   
495 2020-11-10 10:59:59.999       36378608.11           48025   
496 2020-11-10 11:59:59.999       61214714.46           65654   
497 2020-11-10 12:59:59.999       53310596.35           56464   
498 2020-11-10 13:59:59.999       51393461.99           52277   
499 2020-11-10 14:59:59.999       13472265.85           12280   

     takerBuyBaseAssetVolume  takerBuyQuoteAssetVolume  ignore    DEMA_9  \
0                    1126.58               13442658.74       0       NaN   
1                    1251.68               14872868.67       0       NaN   
2                     642.19                7647593.84       0       NaN   
3                     603.06                7199003.46       0       NaN   
4                     757.03                9026310.96       0       NaN   
..                       ...                       ...     ...       ...   
495                  1116.96               17178903.67       0  15335.93   
496                  1954.02               29797695.54       0  15315.92   
497                  1696.47               25952126.02       0  15304.66   
498                  1412.22               21534514.11       0  15303.70   
499                   391.22                5974532.39       0  15278.96   

       SMA_21  RSI_14  MFI_14  
0         NaN     NaN     NaN  
1         NaN    0.00     NaN  
2         NaN   64.75     NaN  
3         NaN   80.38     NaN  
4         NaN   47.55     NaN  
..        ...     ...     ...  
495  15292.98   48.51   43.89  
496  15301.46   47.16   42.28  
497  15315.11   47.67   50.36  
498  15324.41   48.88   43.06  
499  15325.10   44.96   45.31  

[500 rows x 16 columns]

In [10]:
df[['# openTime', 'open', 'high', 'low', 'close', 'volume', 'DEMA_9', 'SMA_21', 'RSI_14',
       'MFI_14']].tail(10).round(2)

# openTime      open      high       low     close   volume  \
490 2020-11-10 05:00:00  15387.74  15441.18  15308.21  15357.92  2178.82   
491 2020-11-10 06:00:00  15357.92  15390.18  15247.21  15275.82  2558.92   
492 2020-11-10 07:00:00  15275.83  15333.49  15250.47  15274.74  1981.50   
493 2020-11-10 08:00:00  15274.75  15460.00  15261.01  15371.37  3795.22   
494 2020-11-10 09:00:00  15371.36  15440.00  15367.40  15383.84  2253.77   
495 2020-11-10 10:00:00  15383.84  15430.00  15302.58  15305.14  2365.38   
496 2020-11-10 11:00:00  15305.15  15323.67  15155.00  15279.24  4014.34   
497 2020-11-10 12:00:00  15279.24  15393.65  15213.00  15287.61  3484.42   
498 2020-11-10 13:00:00  15287.61  15330.00  15179.00  15306.83  3371.00   
499 2020-11-10 14:00:00  15306.84  15308.04  15238.01  15239.49   882.07   

       DEMA_9    SMA_21  RSI_14  MFI_14  
490  15358.95  15336.09   51.12   54.83  
491  15331.42  15329.98   47.23   63.64  
492  15311.19  15320.88   47.18   56.00  
493  15331.57  15300.35   52.14   56.01  
494  15350.69  15291.71   52.76   52.94  
495  15335.93  15292.98   48.51   43.89  
496  15315.92  15301.46   47.16   42.28  
497  15304.66  15315.11   47.67   50.36  
498  15303.70  15324.41   48.88   43.06  
499  15278.96  15325.10   44.96   45.31

Visuele controle van de waarden in Tradingview geven aan dat waardes kloppen maar de tijd 1 uur verschilt t.o.v. UTC+1 time.

## Berekenen van moving averages op RSI14 en MFI14

In [11]:
df['RSI_SMA_14'] = df['RSI_14'].rolling(window=14).mean()
df['MFI_SMA_21'] = df['MFI_14'].rolling(window=21).mean()

In [12]:
df.columns

Index(['# openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime',
       'qouteAssetVolume', 'numberOfTrades', 'takerBuyBaseAssetVolume',
       'takerBuyQuoteAssetVolume', 'ignore', 'DEMA_9', 'SMA_21', 'RSI_14',
       'MFI_14', 'RSI_SMA_14', 'MFI_SMA_21'],
      dtype='object')

In [13]:
df[['# openTime','RSI_14', 'RSI_SMA_14','MFI_14', 'MFI_SMA_21']].tail(10).round(2)

# openTime  RSI_14  RSI_SMA_14  MFI_14  MFI_SMA_21
490 2020-11-10 05:00:00   51.12       49.75   54.83       35.97
491 2020-11-10 06:00:00   47.23       50.07   63.64       36.86
492 2020-11-10 07:00:00   47.18       50.11   56.00       37.35
493 2020-11-10 08:00:00   52.14       50.27   56.01       37.29
494 2020-11-10 09:00:00   52.76       50.38   52.94       37.71
495 2020-11-10 10:00:00   48.51       50.15   43.89       38.21
496 2020-11-10 11:00:00   47.16       49.84   42.28       38.95
497 2020-11-10 12:00:00   47.67       49.91   50.36       40.33
498 2020-11-10 13:00:00   48.88       49.82   43.06       41.45
499 2020-11-10 14:00:00   44.96       49.53   45.31       42.44

## Kolommen controleren op long (1) of short (0) condities

Elke vergelijking krijgt een uitkomst mee:

* Als DEMA9 > SMA21 = 1
* Als RSI14 > RSI_MA_14 = 1
* Als MFI14 > MFI_MA_21 = 1

Als waarde is 0, dan is dit een SHORT conditie.  
Als waarde is 1 of 2, dan is dit een NEUTRAAL conditie ( = huidige positie houden en opletten).  
Als waarde is 3, dan is dit een LONG conditie.  

In [14]:
# SMA check
df['MA_pos'] = np.where(df['DEMA_9']>df['SMA_21'], 1 , 0)
# RSI check
df['RSI_pos'] = np.where(df['RSI_14']>df['RSI_SMA_14'], 1 , 0)
# MFI check
df['MFI_pos'] = np.where(df['MFI_14']>df['MFI_SMA_21'], 1 , 0)

In [15]:
df[['# openTime','close','DEMA_9', 'SMA_21','MA_pos', 'RSI_14','RSI_SMA_14','RSI_pos','MFI_14', 'MFI_SMA_21', 'MFI_pos']].tail(15).round(2)

# openTime     close    DEMA_9    SMA_21  MA_pos  RSI_14  \
485 2020-11-10 00:00:00  15300.00  15298.36  15341.25       0   48.99   
486 2020-11-10 01:00:00  15392.01  15330.65  15341.71       0   52.67   
487 2020-11-10 02:00:00  15339.14  15334.74  15335.69       0   50.42   
488 2020-11-10 03:00:00  15334.91  15336.01  15333.77       1   50.24   
489 2020-11-10 04:00:00  15387.50  15355.72  15337.79       1   52.56   
490 2020-11-10 05:00:00  15357.92  15358.95  15336.09       1   51.12   
491 2020-11-10 06:00:00  15275.82  15331.42  15329.98       1   47.23   
492 2020-11-10 07:00:00  15274.74  15311.19  15320.88       0   47.18   
493 2020-11-10 08:00:00  15371.37  15331.57  15300.35       1   52.14   
494 2020-11-10 09:00:00  15383.84  15350.69  15291.71       1   52.76   
495 2020-11-10 10:00:00  15305.14  15335.93  15292.98       1   48.51   
496 2020-11-10 11:00:00  15279.24  15315.92  15301.46       1   47.16   
497 2020-11-10 12:00:00  15287.61  15304.66  15315.11       0   47.67   
498 2020-11-10 13:00:00  15306.83  15303.70  15324.41       0   48.88   
499 2020-11-10 14:00:00  15239.49  15278.96  15325.10       0   44.96   

     RSI_SMA_14  RSI_pos  MFI_14  MFI_SMA_21  MFI_pos  
485       49.23        0   32.33       39.33        0  
486       48.13        1   27.96       37.45        0  
487       47.71        1   34.39       35.96        0  
488       48.01        1   39.49       35.00        1  
489       48.84        1   47.72       35.00        1  
490       49.75        1   54.83       35.97        1  
491       50.07        0   63.64       36.86        1  
492       50.11        0   56.00       37.35        1  
493       50.27        1   56.01       37.29        1  
494       50.38        1   52.94       37.71        1  
495       50.15        0   43.89       38.21        1  
496       49.84        0   42.28       38.95        1  
497       49.91        0   50.36       40.33        1  
498       49.82        0   43.06       41.45        1  
499       49.53        0   45.31       42.44        1

Nu een kolom voor de telling van de posities en een kolom die obv. deze telling een advies geeft.

In [16]:
# Posities tellen
df['POS_cnt'] = df['MA_pos'] + df['RSI_pos'] + df['MFI_pos']
df[['MA_pos', 'RSI_pos', 'MFI_pos', 'POS_cnt']].tail(15)

MA_pos  RSI_pos  MFI_pos  POS_cnt
485       0        0        0        0
486       0        1        0        1
487       0        1        0        1
488       1        1        1        3
489       1        1        1        3
490       1        1        1        3
491       1        0        1        2
492       0        0        1        1
493       1        1        1        3
494       1        1        1        3
495       1        0        1        2
496       1        0        1        2
497       0        0        1        1
498       0        0        1        1
499       0        0        1        1

In [17]:
df.loc[df['POS_cnt'] == 3, ['# openTime','close']]

# openTime     close
44  2020-10-22 15:00:00  12930.00
45  2020-10-22 16:00:00  13041.75
46  2020-10-22 17:00:00  13046.96
47  2020-10-22 18:00:00  13042.48
48  2020-10-22 19:00:00  13072.48
..                  ...       ...
488 2020-11-10 03:00:00  15334.91
489 2020-11-10 04:00:00  15387.50
490 2020-11-10 05:00:00  15357.92
493 2020-11-10 08:00:00  15371.37
494 2020-11-10 09:00:00  15383.84

[155 rows x 2 columns]

In [18]:
# Advies geven
def f(row):
    if row['POS_cnt'] == 1:
        val = 'NEUTRAAL'
    elif row['POS_cnt'] == 2:
        val = 'NEUTRAAL'
    elif row['POS_cnt'] == 3:
        val = 'LONG'
    else:
        val = 'SHORT'
    return val

#create new column 'POS-adv' using the function above
df['POS_adv'] = df.apply(f, axis=1)

In [19]:
df[['# openTime','close','POS_cnt','POS_adv']].tail(25)

# openTime     close  POS_cnt   POS_adv
475 2020-11-09 14:00:00  15101.00        0     SHORT
476 2020-11-09 15:00:00  15000.96        0     SHORT
477 2020-11-09 16:00:00  15111.62        0     SHORT
478 2020-11-09 17:00:00  15225.00        0     SHORT
479 2020-11-09 18:00:00  15317.51        0     SHORT
480 2020-11-09 19:00:00  15360.00        1  NEUTRAAL
481 2020-11-09 20:00:00  15373.76        1  NEUTRAAL
482 2020-11-09 21:00:00  15369.24        1  NEUTRAAL
483 2020-11-09 22:00:00  15242.61        0     SHORT
484 2020-11-09 23:00:00  15328.41        1  NEUTRAAL
485 2020-11-10 00:00:00  15300.00        0     SHORT
486 2020-11-10 01:00:00  15392.01        1  NEUTRAAL
487 2020-11-10 02:00:00  15339.14        1  NEUTRAAL
488 2020-11-10 03:00:00  15334.91        3      LONG
489 2020-11-10 04:00:00  15387.50        3      LONG
490 2020-11-10 05:00:00  15357.92        3      LONG
491 2020-11-10 06:00:00  15275.82        2  NEUTRAAL
492 2020-11-10 07:00:00  15274.74        1  NEUTRAAL
493 2020-11-10 08:00:00  15371.37        3      LONG
494 2020-11-10 09:00:00  15383.84        3      LONG
495 2020-11-10 10:00:00  15305.14        2  NEUTRAAL
496 2020-11-10 11:00:00  15279.24        2  NEUTRAAL
497 2020-11-10 12:00:00  15287.61        1  NEUTRAAL
498 2020-11-10 13:00:00  15306.83        1  NEUTRAAL
499 2020-11-10 14:00:00  15239.49        1  NEUTRAAL

In [20]:
df.loc[df['POS_adv'] == 'LONG', ['POS_adv','# openTime','close']].tail(25)

POS_adv          # openTime     close
379    LONG 2020-11-05 14:00:00  14894.46
380    LONG 2020-11-05 15:00:00  15080.26
381    LONG 2020-11-05 16:00:00  15155.87
387    LONG 2020-11-05 22:00:00  15534.98
388    LONG 2020-11-05 23:00:00  15590.02
389    LONG 2020-11-06 00:00:00  15864.51
447    LONG 2020-11-08 10:00:00  15161.50
448    LONG 2020-11-08 11:00:00  15171.00
449    LONG 2020-11-08 12:00:00  15242.07
450    LONG 2020-11-08 13:00:00  15204.80
451    LONG 2020-11-08 14:00:00  15343.83
452    LONG 2020-11-08 15:00:00  15353.48
453    LONG 2020-11-08 16:00:00  15387.99
454    LONG 2020-11-08 17:00:00  15574.19
455    LONG 2020-11-08 18:00:00  15528.01
456    LONG 2020-11-08 19:00:00  15339.47
457    LONG 2020-11-08 20:00:00  15375.30
459    LONG 2020-11-08 22:00:00  15500.00
460    LONG 2020-11-08 23:00:00  15475.10
461    LONG 2020-11-09 00:00:00  15549.62
488    LONG 2020-11-10 03:00:00  15334.91
489    LONG 2020-11-10 04:00:00  15387.50
490    LONG 2020-11-10 05:00:00  15357.92
493    LONG 2020-11-10 08:00:00  15371.37
494    LONG 2020-11-10 09:00:00  15383.84

Print nu alleen de waarde van de positie die ingenomen dient te worden van de laatste volle dag = 498 (regel 499 is altijd van de lopende dag en derhalve nog niet afgesloten).

In [21]:
print (df[['# openTime','POS_adv']].iloc[-2])

# openTime    2020-11-10 13:00:00
POS_adv                  NEUTRAAL
Name: 498, dtype: object


In [22]:
advice = df.iloc[-2]['POS_adv']
date = df.iloc[-2]['# openTime']
print(date)
print(advice)

2020-11-10 13:00:00
NEUTRAAL


## RSI oversold / overbought indicator toevoegen

- Oversold conditie is als RSI14 < 30
- Overbought conditie als RSI14 > 70

In [23]:
df['RSI_oversold'] = df['RSI_14'] < 30
df['RSI_overbought'] = df['RSI_14'] > 70

df[['# openTime','RSI_14','RSI_oversold', 'RSI_overbought']].tail(5)

# openTime     RSI_14  RSI_oversold  RSI_overbought
495 2020-11-10 10:00:00  48.507678         False           False
496 2020-11-10 11:00:00  47.161232         False           False
497 2020-11-10 12:00:00  47.666785         False           False
498 2020-11-10 13:00:00  48.876401         False           False
499 2020-11-10 14:00:00  44.955736         False           False

In [24]:
# Oversold data zoeken
df.loc[df['RSI_oversold'] == True, ['# openTime','RSI_14']].tail(25)

# openTime     RSI_14
1   2020-10-20 20:00:00   0.000000
186 2020-10-28 13:00:00  28.433393
187 2020-10-28 14:00:00  27.321446
301 2020-11-02 08:00:00  29.135474
305 2020-11-02 12:00:00  28.696957
431 2020-11-07 18:00:00  28.241950
432 2020-11-07 19:00:00  22.384403

In [25]:
# Overbought data zoeken
df.loc[df['RSI_overbought'] == True, ['# openTime','RSI_14']].tail(25)

# openTime     RSI_14
175 2020-10-28 02:00:00  74.360959
340 2020-11-03 23:00:00  74.054683
368 2020-11-05 03:00:00  72.241104
369 2020-11-05 04:00:00  70.390756
370 2020-11-05 05:00:00  71.395017
371 2020-11-05 06:00:00  72.138259
372 2020-11-05 07:00:00  77.033408
373 2020-11-05 08:00:00  73.875960
374 2020-11-05 09:00:00  74.522828
375 2020-11-05 10:00:00  79.951696
376 2020-11-05 11:00:00  81.540006
377 2020-11-05 12:00:00  84.448055
378 2020-11-05 13:00:00  79.334902
379 2020-11-05 14:00:00  81.085796
380 2020-11-05 15:00:00  84.317470
381 2020-11-05 16:00:00  85.409952
382 2020-11-05 17:00:00  76.443910
384 2020-11-05 19:00:00  74.257006
385 2020-11-05 20:00:00  70.926274
386 2020-11-05 21:00:00  74.563661
387 2020-11-05 22:00:00  79.781171
388 2020-11-05 23:00:00  80.549694
389 2020-11-06 00:00:00  83.847172
390 2020-11-06 01:00:00  78.264582
391 2020-11-06 02:00:00  72.775402

Interessant om te weten wat de verhoudingen zijn tussen oversold en overbought. Kan aangeven dat je minder kansen krijgt om op een dip te kopen dan op overbought te verkopen.

## Toevoegen van ATR indicator

In [26]:
df.ta.atr(length=14, append=True)

df.columns

Index(['# openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime',
       'qouteAssetVolume', 'numberOfTrades', 'takerBuyBaseAssetVolume',
       'takerBuyQuoteAssetVolume', 'ignore', 'DEMA_9', 'SMA_21', 'RSI_14',
       'MFI_14', 'RSI_SMA_14', 'MFI_SMA_21', 'MA_pos', 'RSI_pos', 'MFI_pos',
       'POS_cnt', 'POS_adv', 'RSI_oversold', 'RSI_overbought', 'ATR_14'],
      dtype='object')

In [27]:
df[['# openTime', 'ATR_14']].tail(5)

# openTime      ATR_14
495 2020-11-10 10:00:00  159.098814
496 2020-11-10 11:00:00  159.782470
497 2020-11-10 12:00:00  161.273008
498 2020-11-10 13:00:00  160.539221
499 2020-11-10 14:00:00  154.074277

ATR bedragen kloppen met de Tradingview chart!

In [28]:
# Adding ATR and SL/TP1 levels
df['Close+atr']=df['close'] + df['ATR_14']
df['Close-atr']=df['close'] - df['ATR_14']
df[['# openTime','close','ATR_14','Close+atr','Close-atr']].tail(10)

# openTime     close      ATR_14     Close+atr     Close-atr
490 2020-11-10 05:00:00  15357.92  174.671126  15532.591126  15183.248874
491 2020-11-10 06:00:00  15275.82  172.406760  15448.226760  15103.413240
492 2020-11-10 07:00:00  15274.74  166.021991  15440.761991  15108.718009
493 2020-11-10 08:00:00  15371.37  168.376849  15539.746849  15202.993151
494 2020-11-10 09:00:00  15383.84  161.535645  15545.375645  15222.304355
495 2020-11-10 10:00:00  15305.14  159.098814  15464.238814  15146.041186
496 2020-11-10 11:00:00  15279.24  159.782470  15439.022470  15119.457530
497 2020-11-10 12:00:00  15287.61  161.273008  15448.883008  15126.336992
498 2020-11-10 13:00:00  15306.83  160.539221  15467.369221  15146.290779
499 2020-11-10 14:00:00  15239.49  154.074277  15393.564277  15085.415723

# Alle kolommen

In [29]:
df.columns

Index(['# openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime',
       'qouteAssetVolume', 'numberOfTrades', 'takerBuyBaseAssetVolume',
       'takerBuyQuoteAssetVolume', 'ignore', 'DEMA_9', 'SMA_21', 'RSI_14',
       'MFI_14', 'RSI_SMA_14', 'MFI_SMA_21', 'MA_pos', 'RSI_pos', 'MFI_pos',
       'POS_cnt', 'POS_adv', 'RSI_oversold', 'RSI_overbought', 'ATR_14',
       'Close+atr', 'Close-atr'],
      dtype='object')

In [30]:
df[['# openTime', 'open', 'high', 'low', 'close', 'volume', 'closeTime',
        'DEMA_9', 'SMA_21', 'RSI_14',
       'MFI_14', 'RSI_SMA_14', 'MFI_SMA_21', 'MA_pos', 'RSI_pos', 'MFI_pos',
       'POS_cnt', 'POS_adv', 'RSI_oversold', 'ATR_14', 'Close+atr',
       'Close-atr','RSI_overbought']].tail(10).round(2)

# openTime      open      high       low     close   volume  \
490 2020-11-10 05:00:00  15387.74  15441.18  15308.21  15357.92  2178.82   
491 2020-11-10 06:00:00  15357.92  15390.18  15247.21  15275.82  2558.92   
492 2020-11-10 07:00:00  15275.83  15333.49  15250.47  15274.74  1981.50   
493 2020-11-10 08:00:00  15274.75  15460.00  15261.01  15371.37  3795.22   
494 2020-11-10 09:00:00  15371.36  15440.00  15367.40  15383.84  2253.77   
495 2020-11-10 10:00:00  15383.84  15430.00  15302.58  15305.14  2365.38   
496 2020-11-10 11:00:00  15305.15  15323.67  15155.00  15279.24  4014.34   
497 2020-11-10 12:00:00  15279.24  15393.65  15213.00  15287.61  3484.42   
498 2020-11-10 13:00:00  15287.61  15330.00  15179.00  15306.83  3371.00   
499 2020-11-10 14:00:00  15306.84  15308.04  15238.01  15239.49   882.07   

                  closeTime    DEMA_9    SMA_21  RSI_14  ...  MA_pos  RSI_pos  \
490 2020-11-10 05:59:59.999  15358.95  15336.09   51.12  ...       1        1   
491 2020-11-10 06:59:59.999  15331.42  15329.98   47.23  ...       1        0   
492 2020-11-10 07:59:59.999  15311.19  15320.88   47.18  ...       0        0   
493 2020-11-10 08:59:59.999  15331.57  15300.35   52.14  ...       1        1   
494 2020-11-10 09:59:59.999  15350.69  15291.71   52.76  ...       1        1   
495 2020-11-10 10:59:59.999  15335.93  15292.98   48.51  ...       1        0   
496 2020-11-10 11:59:59.999  15315.92  15301.46   47.16  ...       1        0   
497 2020-11-10 12:59:59.999  15304.66  15315.11   47.67  ...       0        0   
498 2020-11-10 13:59:59.999  15303.70  15324.41   48.88  ...       0        0   
499 2020-11-10 14:59:59.999  15278.96  15325.10   44.96  ...       0        0   

     MFI_pos  POS_cnt   POS_adv  RSI_oversold  ATR_14 Close+atr  Close-atr  \
490        1        3      LONG         False  174.67  15532.59   15183.25   
491        1        2  NEUTRAAL         False  172.41  15448.23   15103.41   
492        1        1  NEUTRAAL         False  166.02  15440.76   15108.72   
493        1        3      LONG         False  168.38  15539.75   15202.99   
494        1        3      LONG         False  161.54  15545.38   15222.30   
495        1        2  NEUTRAAL         False  159.10  15464.24   15146.04   
496        1        2  NEUTRAAL         False  159.78  15439.02   15119.46   
497        1        1  NEUTRAAL         False  161.27  15448.88   15126.34   
498        1        1  NEUTRAAL         False  160.54  15467.37   15146.29   
499        1        1  NEUTRAAL         False  154.07  15393.56   15085.42   

     RSI_overbought  
490           False  
491           False  
492           False  
493           False  
494           False  
495           False  
496           False  
497           False  
498           False  
499           False  

[10 rows x 23 columns]

In [48]:
# Bepaal de laagste RSI stand (oversold) onder short condities. Kans op koersstijding is hier aanwezig (na double dip)
df.loc[(df.RSI_oversold == True) & (df.POS_adv == 'SHORT') , ['# openTime']].tail(25)

# openTime
1   2020-10-20 20:00:00
186 2020-10-28 13:00:00
187 2020-10-28 14:00:00
301 2020-11-02 08:00:00
305 2020-11-02 12:00:00
431 2020-11-07 18:00:00
432 2020-11-07 19:00:00

In [32]:
# Bepaal de hoogste RSI stand (overbought) onder LONG condities. Kans op koersdaling is hier aanwezig (na double top)
df.loc[(df.RSI_overbought == True) & (df.POS_adv == 'LONG') , ['# openTime']].tail(25)


# openTime
164 2020-10-27 15:00:00
165 2020-10-27 16:00:00
166 2020-10-27 17:00:00
167 2020-10-27 18:00:00
168 2020-10-27 19:00:00
170 2020-10-27 21:00:00
171 2020-10-27 22:00:00
340 2020-11-03 23:00:00
368 2020-11-05 03:00:00
369 2020-11-05 04:00:00
370 2020-11-05 05:00:00
371 2020-11-05 06:00:00
372 2020-11-05 07:00:00
373 2020-11-05 08:00:00
374 2020-11-05 09:00:00
375 2020-11-05 10:00:00
376 2020-11-05 11:00:00
377 2020-11-05 12:00:00
378 2020-11-05 13:00:00
379 2020-11-05 14:00:00
380 2020-11-05 15:00:00
381 2020-11-05 16:00:00
387 2020-11-05 22:00:00
388 2020-11-05 23:00:00
389 2020-11-06 00:00:00

## Nieuw idee: Check in Panda's of een veldwaarde anders is t.o.v. de vorige waarde

Dus als veld 499 != 498 , dan 1 anders 0.  <
Als 1 dan waarschuwing via signals

In [35]:
df.iloc[-2]['POS_adv']

'NEUTRAAL'

In [55]:
df['POS_adv_changed'] = df['POS_adv'].shift(+1)==df['POS_adv']
df[['POS_adv_changed','POS_adv','# openTime']].tail(30)

POS_adv_changed   POS_adv          # openTime
470             True     SHORT 2020-11-09 09:00:00
471            False  NEUTRAAL 2020-11-09 10:00:00
472             True  NEUTRAAL 2020-11-09 11:00:00
473             True  NEUTRAAL 2020-11-09 12:00:00
474             True  NEUTRAAL 2020-11-09 13:00:00
475            False     SHORT 2020-11-09 14:00:00
476             True     SHORT 2020-11-09 15:00:00
477             True     SHORT 2020-11-09 16:00:00
478             True     SHORT 2020-11-09 17:00:00
479             True     SHORT 2020-11-09 18:00:00
480            False  NEUTRAAL 2020-11-09 19:00:00
481             True  NEUTRAAL 2020-11-09 20:00:00
482             True  NEUTRAAL 2020-11-09 21:00:00
483            False     SHORT 2020-11-09 22:00:00
484            False  NEUTRAAL 2020-11-09 23:00:00
485            False     SHORT 2020-11-10 00:00:00
486            False  NEUTRAAL 2020-11-10 01:00:00
487             True  NEUTRAAL 2020-11-10 02:00:00
488            False      LONG 2020-11-10 03:00:00
489             True      LONG 2020-11-10 04:00:00
490             True      LONG 2020-11-10 05:00:00
491            False  NEUTRAAL 2020-11-10 06:00:00
492             True  NEUTRAAL 2020-11-10 07:00:00
493            False      LONG 2020-11-10 08:00:00
494             True      LONG 2020-11-10 09:00:00
495            False  NEUTRAAL 2020-11-10 10:00:00
496             True  NEUTRAAL 2020-11-10 11:00:00
497             True  NEUTRAAL 2020-11-10 12:00:00
498             True  NEUTRAAL 2020-11-10 13:00:00
499             True  NEUTRAAL 2020-11-10 14:00:00

In [53]:
df.loc[df.POS_adv_changed== False , ['# openTime']].tail(25)

# openTime
382 2020-11-05 17:00:00
387 2020-11-05 22:00:00
390 2020-11-06 01:00:00
404 2020-11-06 15:00:00
411 2020-11-06 22:00:00
420 2020-11-07 07:00:00
421 2020-11-07 08:00:00
426 2020-11-07 13:00:00
438 2020-11-08 01:00:00
447 2020-11-08 10:00:00
458 2020-11-08 21:00:00
459 2020-11-08 22:00:00
462 2020-11-09 01:00:00
468 2020-11-09 07:00:00
471 2020-11-09 10:00:00
475 2020-11-09 14:00:00
480 2020-11-09 19:00:00
483 2020-11-09 22:00:00
484 2020-11-09 23:00:00
485 2020-11-10 00:00:00
486 2020-11-10 01:00:00
488 2020-11-10 03:00:00
491 2020-11-10 06:00:00
493 2020-11-10 08:00:00
495 2020-11-10 10:00:00

In [56]:
df

# openTime      open      high       low     close       volume  \
0   2020-10-20 19:00:00  11964.58  11981.20  11881.00  11912.35  2744.144578   
1   2020-10-20 20:00:00  11912.42  11912.79  11855.00  11897.92  2523.731342   
2   2020-10-20 21:00:00  11897.92  11937.73  11877.49  11922.53  1409.942909   
3   2020-10-20 22:00:00  11922.53  11961.00  11918.22  11950.66  1270.092250   
4   2020-10-20 23:00:00  11950.66  11961.16  11885.00  11909.99  1551.691786   
..                  ...       ...       ...       ...       ...          ...   
495 2020-11-10 10:00:00  15383.84  15430.00  15302.58  15305.14  2365.383922   
496 2020-11-10 11:00:00  15305.15  15323.67  15155.00  15279.24  4014.340547   
497 2020-11-10 12:00:00  15279.24  15393.65  15213.00  15287.61  3484.423100   
498 2020-11-10 13:00:00  15287.61  15330.00  15179.00  15306.83  3370.999279   
499 2020-11-10 14:00:00  15306.84  15308.04  15238.01  15239.49   882.070653   

                  closeTime  qouteAssetVolume  numberOfTrades  \
0   2020-10-20 19:59:59.999      3.274267e+07           39342   
1   2020-10-20 20:59:59.999      2.998975e+07           39583   
2   2020-10-20 21:59:59.999      1.679021e+07           28867   
3   2020-10-20 22:59:59.999      1.516067e+07           27532   
4   2020-10-20 23:59:59.999      1.850040e+07           24948   
..                      ...               ...             ...   
495 2020-11-10 10:59:59.999      3.637861e+07           48025   
496 2020-11-10 11:59:59.999      6.121471e+07           65654   
497 2020-11-10 12:59:59.999      5.331060e+07           56464   
498 2020-11-10 13:59:59.999      5.139346e+07           52277   
499 2020-11-10 14:59:59.999      1.347227e+07           12280   

     takerBuyBaseAssetVolume  ...  RSI_pos  MFI_pos  POS_cnt   POS_adv  \
0                1126.575869  ...        0        0        0     SHORT   
1                1251.684940  ...        0        0        0     SHORT   
2                 642.188103  ...        0        0        0     SHORT   
3                 603.063086  ...        0        0        0     SHORT   
4                 757.031464  ...        0        0        0     SHORT   
..                       ...  ...      ...      ...      ...       ...   
495              1116.957106  ...        0        1        2  NEUTRAAL   
496              1954.016547  ...        0        1        2  NEUTRAAL   
497              1696.465821  ...        0        1        1  NEUTRAAL   
498              1412.224206  ...        0        1        1  NEUTRAAL   
499               391.220254  ...        0        1        1  NEUTRAAL   

     RSI_oversold  RSI_overbought      ATR_14     Close+atr     Close-atr  \
0           False           False  100.200000  12012.550000  11812.150000   
1            True           False   78.209630  11976.129630  11819.710370   
2           False           False   71.770786  11994.300786  11850.759214   
3           False            True   63.698669  12014.358669  11886.961331   
4           False           False   66.573296  11976.563296  11843.416704   
..            ...             ...         ...           ...           ...   
495         False           False  159.098814  15464.238814  15146.041186   
496         False           False  159.782470  15439.022470  15119.457530   
497         False           False  161.273008  15448.883008  15126.336992   
498         False           False  160.539221  15467.369221  15146.290779   
499         False           False  154.074277  15393.564277  15085.415723   

     POS_adv_changed  
0              False  
1               True  
2               True  
3               True  
4               True  
..               ...  
495            False  
496             True  
497             True  
498             True  
499             True  

[500 rows x 29 columns]